### ISSUE TITLE
Add tests for unit conversion

### ISSUE THREAD

**Issue by MichaelClerx** [ModellingWebLab/weblab-fc#122]
Done:

- Time var unit conversion (graph state)
- State var unit conversion (graph state)
- Output unit conversion (graph state)

To-do:
- Input unit conversion: Changing initial value on existing state
- Input unit conversion: Changing initial value on constant
- Input unit conversion: Using define to change a constant
- Input unit conversion: Using define to change an ODE

**Comment by MichaelClerx** [comment]
Thinking we could:

1. Have a protocol that sets initial conditions for V in V, and then see if the output trace is correct (for an AP model this would test that it still interacts with the currents in mV, or the AP would be nuts)
2. Run `test_sim_environments` on a model in V to test the case where V is also redefined
3. Have a protocol that sets some constant, e.g. `gNa` in an unusual unit, just check V output to see if it worked
4. Same, but now with define

where 1, 3, and 4 could all be the same protocol?


Thinking I could write a model that looks like this:

```
dot(V) = -1/C * i_ion
    in [mV]
V2 = V
    in [mV]
dot(K) = 0.1
    in [mM]
K2 = K
    in [mM]
gNa = 1 [nA]
gK = 1 [pA]
```
and then add a protocol that
- Sets V in V, but reads V2 in mM

**Comment by MichaelClerx** [comment]
I'm not sure about 2, because if it doesn't convert the units we still get the expected output, I think

**Comment by jonc125** [comment]
Yeah, this looks reasonable. A few thoughts...

1. For this I'd use the `test/models/single_ode.cellml` test model, as it's easy to tell analytically what the result should be (and quicker to process). But you're right that separating the testing of changing initial value for an existing state variable from the case where we `define` a new ODE is sensible.
2. Yep, can use the paci model (as in one of the graphstate tests). Maybe check first that it does break without conversion?
3. The model `test/data/simple_ode.cellml` (NB: different from 1 above!) could be used: it has `dot(V) = a` where `a` is a parameter.
4. This could use the same protocol & model as (3), but alter the `dot(Na) = n` equation?

**Comment by jonc125** [comment]
I suspect that if you got (1) to pass then this would force (2) to fail if the conversion hadn't been done right - it'd be very hard to get a false positive.

**Comment by MichaelClerx** [comment]
I don't quite understand how we'd know if it went wrong. I can accidentally set V to mega amps, then run a simulation, and that wouldn't change the numbers returned by the simulation, right?

That's why I was thinking we needed a model where something depends on V (and V depends on that something), so that it only passes if the conversion has actually been applied?

**Comment by jonc125** [comment]
Well, the single_ode model has `dot(V) = 1 mV/ms` with `V(0)=2 mV`. If you have `V` set as an input & output in volts in the protocol, and override `V(0) = 0.001 V`, then if it's converted you should read out 0.001, 0.002, etc. If it wasn't you'd probably get 0.001, 1.001, 2.001, etc?

**Comment by MichaelClerx** [comment]
Oh ok! Because we don't change the RHS. Thanks!

**Comment by jonc125** [comment]
We could also test with a variant of the graph state protocol that allows you to set the initial V as well as just read out the state, for extra safety.

**Comment by MichaelClerx** [comment]
`simple_ode.cellml` uses annotations like `urn:test-ns#parameter_a`. Can I access those yet, or will I have to implement that @jonc125 ?

**Comment by jonc125** [comment]
Yep, you just need to bind the namespace to a prefix in your protocol, e.g.
```
namespace test = "urn:test-ns#"
```
Then use `test:parameter_a`.

**Comment by MichaelClerx** [comment]
Closed in #132
